### PR TITLE
Add support for toolchain compilation with LLVM_LINK_LLVM_DYLIB option

### PR DESCRIPTION
**Pull request by fineg74**
This is to enable compilation of the toolchain with dynamic libraries

### CODE REVIEW

File: llvm-spirv/lib/SPIRV/CMakeLists.txt [hunk 1]
```diff
@@ -1,54 +1,107 @@
...
-    Core
-    Demangle
-    IRReader
-    Linker
-    Passes
-    Support
-    TransformUtils
+if(LLVM_LINK_LLVM_DYLIB)
+  add_llvm_library(LLVMSPIRVLib STATIC DISABLE_LLVM_LINK_LLVM_DYLIB
+    LLVMSPIRVOpts.cpp
```
Comment (v-klochkov) [inline llvm-spirv/lib/SPIRV/CMakeLists.txt:3]: This list of 36 files is copied twice.
Please create a variable for them, then reuse it in 2 places like it is done here: https://github.com/intel/llvm/blob/sycl/sycl/source/CMakeLists.txt#L116

**Comment by v-klochkov** [comment]
If you add additional fix for this comment https://github.com/intel/llvm/pull/6413#discussion_r916425618 then the fix is good, IMO. 
I would feel safer though if one more reviewer could take a look too, perhaps @bader or @steffenlarsen

File: llvm-spirv/lib/SPIRV/CMakeLists.txt [hunk 2]
```diff
@@ -1,54 +1,107 @@
...
-    BitWriter
-    CodeGen
-    Core
-    Demangle
-    IRReader
-    Linker
-    Passes
-    Support
-    TransformUtils
+if(LLVM_LINK_LLVM_DYLIB)
```
Comment (bader) [inline llvm-spirv/lib/SPIRV/CMakeLists.txt:1]: @fineg74, please, commit LLVM-SPIRV-Translator changes to https://github.com/KhronosGroup/SPIRV-LLVM-Translator/ repository. We regularly pull changes from Khronos repository here.

I thought that llvm-spirv tool build is fixed by https://github.com/KhronosGroup/SPIRV-LLVM-Translator/pull/458. Has it been broken again?

File: llvm/tools/spirv-to-ir-wrapper/CMakeLists.txt [hunk 3]
```diff
@@ -1,14 +1,37 @@
...
+  )
+else()
+  set(LLVM_LINK_COMPONENTS
+    BinaryFormat
+    Core
+    Support
+    SPIRVLib
   )
+endif()
+
```
Comment (bader) [inline llvm/tools/spirv-to-ir-wrapper/CMakeLists.txt:16]: Please, remove duplication.
You can do it here by adding to the existing code.
```cmake
if (LLVM_LINK_LLVM_DYLIB)
  list (REMOVE_ITEM LLVM_LINK_COMPONENTS SPIRVLib)
endif()
```
Another option: create a list w/o `SPIRVLib` and add it if `LLVM_LINK_LLVM_DYLIB` is not set.

On the other hand, I expect `add_llvm_tool` to handle `LLVM_LINK_LLVM_DYLIB` correctly. Could you clarify what is the problem with `spirv-to-ir-wrapper` tool and `SPIRVLib` library?

Comment (bader) [reply llvm/tools/spirv-to-ir-wrapper/CMakeLists.txt:11]: I expect GenXIntrinsics library to be linked statically to llvm.so library as all other libraries. Can we do that?

Comment (bader) [reply llvm/tools/spirv-to-ir-wrapper/CMakeLists.txt:11]: Why did you [remove](https://github.com/intel/vc-intrinsics/commit/c456ff05bd8e11458772acef20e59ef18aa7876d) `GenXIntrinsics` library from `llvm.so` library? I think this is the reason of cycle dependency. All libraries in the error message are part of the llvm.so, except `LLVMGenXIntrinsics`, which also depends on LLVM libraries. That's why we have a conflict: `LLVMSYCLLowerIR` library (part of `llvm.so`) depends on `LLVMGenXIntrinsics` (standalone), which depends on `LLVMCore` library (part of `llvm.so`).
If you make `LLVMGenXIntrinsics` as I suggested above, there should be no cyclic dependencies as there will be just one `llvm.so` library.

File: llvm-spirv/lib/SPIRV/CMakeLists.txt (hunk 2 above)
Comment (pvchupin) [reply llvm-spirv/lib/SPIRV/CMakeLists.txt:39]: It seems https://github.com/KhronosGroup/SPIRV-LLVM-Translator/pull/1543 is the PR mentioned and it should land first.
@fineg74, please add detailed rationale in PR description there. Not sure if testing fails due to infrastructure temporary problems, you can restart if it was. I guess failed testing and no description are the reasons patch stuck for a while.

Comment (bader) [reply llvm-spirv/lib/SPIRV/CMakeLists.txt:39]: https://github.com/KhronosGroup/SPIRV-LLVM-Translator/pull/1543 is merged, so I expect no changes in llvm-spirv/* files.

I agree with Pavel that description should be improved. It's not clear why we need to change customize cmake files here. I expected `add_llvm_library` to handle linking with all intrinsics. I assume we are building all libraries from sources, so compilation and linking with intrinsics_gen should be no different than any other intrinsics like X86 or NVPTX intrinsics, which do not require this changes. Could you elaborate in the description the difference with regular intrinsics, which makes us customize cmake files for these tools, please?

File: llvm/tools/spirv-to-ir-wrapper/CMakeLists.txt (hunk 3 above)
Comment (bader) [reply llvm/tools/spirv-to-ir-wrapper/CMakeLists.txt:11]: > cmake is more sensitive to circular dependencies when dynamic libraries are built

I don't get what it means.
Shouldn't LLVMGenXIntrinsics be a part of LLVMTarget library or some other LLVM target library? Isn't the root case of the problem is that we put intrinsics in a separate library as opposite to how all other intrinsics are handled?

Comment (kbobrovs) [reply llvm/tools/spirv-to-ir-wrapper/CMakeLists.txt:11]: > Could you elaborate in the description the difference with regular intrinsics, which makes us customize cmake files for these tools, please?

GenX intrinsics come from an external repo during the build, unlike others, and they are built with the cmake files provided in the repo.

**Comment by asudarsa** [comment]
Hi @fineg74 

Is this an active PR? Please let me know if I needed to review this. Sorry for delay.

Thanks

**Pull request closed without merging**